---
title: "AMC-Tax_Stats"
output: html_document
date: "2025-12-01"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  warning = FALSE,
  message = FALSE,
  fig.align = "center",
  fig.width = 6,
  fig.height = 4,
  out.width = "80%"
)


```
```

<br><br>

# Kraken2

Taxonomical Analysis

```{r taxonomy_analysis, results="hide"}

# load necessary packages
library(BiocManager)
library(tidyverse)
library(phyloseq)
library(microViz)
library(tidylog)
library(ggpubr)
library(vegan)
library(readxl)
library(viridis)
library(dunn.test)
library(microbiome)
library(pairwiseAdonis)

ps_tax <- readRDS("data/ps_tax.RDS")

```


## Decontam

```{r decontam_tax}

# Decontamination using decontam package
library(decontam)

ps3_tax <- ps_tax %>%
  ps_mutate(
    SampleBinary = if_else(str_detect(Trt, "NegativeControl"), true = "Control", false = "Sample")
  )

# Identify contaminants using prevalence method
sample_data(ps3_tax)$is.neg <- sample_data(ps3_tax)$SampleBinary == "Control"
contamdf.prev <- isContaminant(ps3_tax, method="prevalence", neg="is.neg", threshold = 0.8)

# Create phyloseq object of only contaminants in the controls
con_tax <- rownames(contamdf.prev[contamdf.prev$contaminant == "TRUE", ])
conps_tax <- prune_taxa(rownames(ps3_tax@tax_table) %in% con_tax, ps3_tax)
negps_tax <- subset_samples(conps_tax, is.neg == "TRUE")

# Extract list of contaminants
contams_tax <- rownames(contamdf.prev[contamdf.prev$contaminant == "TRUE", ])
contams_tax # 1000 with 0.8

# Remove contaminant sequences
noncontam_tax <- prune_taxa(!rownames(ps3_tax@tax_table) %in% contams_tax, ps3_tax)

# Filter controls
controls_tax <- ps3_tax %>%
  ps_filter(SampleBinary == "Control", .keep_all_taxa = TRUE)

controls_tax <- sample_data(controls_tax)$Project_ID 

# Remove all controls
noneg_tax <- ps_filter(noncontam_tax, !Project_ID %in% controls_tax)
factor(sample_data(noneg_tax)$Trt)

# Remove positive control
noneg_tax <- ps_filter(noneg_tax, Trt != "PositiveControl")
factor(sample_data(noneg_tax)$Trt)

# Remove weeks 3, 13 and 15
noneg_tax <- ps_filter(noneg_tax, !Age_wk %in% c("3", "13", "15"))

# Prevalence-based filtering for negonly decontam
sort(phyloseq::sample_sums(noneg_tax))
(noneg_tax <- phyloseq::subset_samples(noneg_tax, phyloseq::sample_sums(noneg_tax) > 1))
(noneg_tax <- phyloseq::prune_taxa(phyloseq::taxa_sums(noneg_tax) > 0, noneg_tax))
noneg_tax <- noneg_tax %>% tax_filter(
  min_prevalence = 0.01,
  min_total_abundance = 0.0001)

# Display sample data
sample_data(noneg_tax)

# total % of reads/taxa removed
sum(contamdf.prev$contaminant) / nrow(contamdf.prev)

```

<br>

## Alpha Diversity

```{r alpha_diversity_tax-summary}

# Get alpha diversity metrics
sampdf <- sample_data(noneg_tax) %>%
  data.frame() %>%
  rownames_to_column(var = "ID")

# Calculate alpha diversity
alpha <- estimate_richness(noneg_tax, measures = c("Shannon", "Observed")) %>%
  rownames_to_column(var = "ID") %>%
  merge(sampdf, by = "ID")

# Get summary statistics for Shannon diversity
sum_Shannon <- alpha %>%
  group_by(Trt, Age_wk) %>%
  get_summary_stats(Shannon, type = "mean_sd")

sum_Shannon

# Get summary statistics for Observed diversity
sum_Observed <- alpha %>%
  group_by(Trt, Age_wk) %>%
  get_summary_stats(Observed, type = "mean_sd")

sum_Observed

```

```{r alpha_diversity_tax-outliers}

## 1. Compute alpha diversity first ----
sampdf <- sample_data(noneg_tax) %>%
  data.frame() %>%
  rownames_to_column(var = "ID")

alpha <- estimate_richness(noneg_tax, measures = c("Shannon", "Observed")) %>%
  rownames_to_column(var = "ID") %>%
  left_join(sampdf, by = "ID")

## 2. Summary stats for diagnostics ----
st <- alpha %>%
  summarise(mean = mean(Shannon, na.rm = TRUE),
            sd = sd(Shannon, na.rm = TRUE)) %>%
  mutate(outhi = mean + 3 * sd,
         outlo = mean - 3 * sd)

sto <- alpha %>%
  summarise(mean = mean(Observed, na.rm = TRUE),
            sd = sd(Observed, na.rm = TRUE)) %>%
  mutate(outhi = mean + 3 * sd,
         outlo = mean - 3 * sd)

cat("Shannon mean ± 3SD range:", st$outlo, "to", st$outhi, "\n")
cat("Observed mean ± 3SD range:", sto$outlo, "to", sto$outhi, "\n")

## 3. Detect outliers for each metric ----
detect_outliers <- function(df, metric) {
  stats <- df %>%
    summarise(mean = mean(.data[[metric]], na.rm = TRUE),
              sd = sd(.data[[metric]], na.rm = TRUE))
  upper <- stats$mean + 3 * stats$sd
  lower <- stats$mean - 3 * stats$sd
  df %>%
    filter(.data[[metric]] < lower | .data[[metric]] > upper) %>%
    pull(ID)
}

outlier_shannon  <- detect_outliers(alpha, "Shannon")
outlier_observed <- detect_outliers(alpha, "Observed")

outlier_ids <- unique(c(outlier_shannon, outlier_observed))

cat("Outliers removed:", paste(outlier_ids, collapse = ", "), "\n")

## 4. Filter them out ----
alpha_clean <- alpha %>%
  filter(!ID %in% outlier_ids)

## Optional check
n_removed <- nrow(alpha) - nrow(alpha_clean)
cat("Total removed:", n_removed, "outliers\n")


```

<br>

Shapiro-Wilk Test for Normality

```{r normality_tests_tax}

# Shapiro-Wilk Test for Normality
shapiro_shannon <- shapiro.test(alpha_clean$Shannon)
shapiro_observed <- shapiro.test(alpha_clean$Observed)

shapiro_shannon # W = 0.83512, p-value = 3.118e-06
shapiro_observed # W = 0.66831, p-value = 8.96e-10

```

<br>

[Wilcox Test for Diet Effect]{.underline}

Mann-Whitney and BH adjustment

**Shannon Index**

Taxonomy

```{r Shannon_tax}

## Doing first within Ab and then No - Shannon 
alpha_Ab <- subset(alpha_clean, Diet == "Ab")
alpha_No <- subset(alpha_clean, Diet == "No")

alpha_Ab$Age_wk <- as.numeric(alpha_Ab$Age_wk)
alpha_No$Age_wk <- as.numeric(alpha_No$Age_wk)

kruskal.test(Shannon ~ Age_wk, data = alpha_Ab)  # chi-squared = 0.66667, df = 4, p-value = 0.9554

kruskal.test(Shannon ~ Age_wk, data = alpha_No) #  chi-squared = 3.497, df = 4, p-value = 0.4783

# no pairwise since ns

# linear mixed model
library(lme4)
library(lmerTest)

alpha_clean$Age_wk <- as.numeric(as.character(alpha_clean$Age_wk))

lmm1 <- lmer(Shannon ~ Diet * Age_wk + (1 | Animal_ID), data = alpha_clean)

anova(lmm1)

summary(lmm1)

```

<br>

[Wilcoxon Test for Age Effect]{.underline}

**Observed Index**

Taxonomy

```{r Observed_tax}

## Doing first within Ab and then No - Observed 
alpha_Ab <- subset(alpha_clean, Diet == "Ab")
alpha_No <- subset(alpha_clean, Diet == "No")

alpha_Ab$Age_wk <- as.numeric(alpha_Ab$Age_wk)
alpha_No$Age_wk <- as.numeric(alpha_No$Age_wk)

kruskal.test(Observed ~ Age_wk, data = alpha_Ab) # Kruskal-Wallis chi-squared = 2.1356, df = 4, p-value = 0.7108
kruskal.test(Observed ~ Age_wk, data = alpha_No) # KKruskal-Wallis chi-squared = 1.2327, df = 4, p-value = 0.8727

# no dunn test since ns 

# linear mixed model 
alpha_clean$Age_wk <- as.numeric(as.character(alpha_clean$Age_wk))

lmm_obs <- lmer(Observed ~ Diet * Age_wk + (1 | Animal_ID), data = alpha_clean)

anova(lmm_obs)

summary(lmm_obs)

# plotting
alpha_clean$Diet <- recode_factor(alpha_clean$Diet,
                            "Ab" = "MR+A",
                            "No" = "MR")

treatment_colors <- c("MR+A" = "#B22222", "MR" = "#377EB8")


# plotting 

library(ggpubr)

ggline(alpha_clean, x = "Age_wk", y = "Shannon",
       add = "mean_sd", error.plot = "errorbar",
       color = "Diet", group = "Diet",
       palette = treatment_colors,
       linewidth = 1.0, point.size = 1.5, alpha = 0.8) +
  scale_x_discrete(labels = c("1","5","7","9","11")) +
  labs(x = "Age (weeks)", y = "Shannon Diversity") +
  theme_classic(base_size = 12) +
  theme(
    legend.position = "right",
    axis.title = element_text(size = 12, face = "bold"),
    plot.title = element_blank()   # remove title
  )

ggline(alpha_clean, x = "Age_wk", y = "Observed",
       add = "mean_sd", error.plot = "errorbar",
       color = "Diet", group = "Diet",
       palette = treatment_colors,
       linewidth = 1.0, point.size = 1.5, alpha = 0.8) +
  scale_x_discrete(labels = c("1","5","7","9","11")) +
  scale_y_continuous(expand = c(0,0)) +
  labs(x = "Age (weeks)", y = "Observed Diversity") +
  theme_classic(base_size = 12) +
  theme(
    legend.position = "right",
    axis.title = element_text(size = 12, face = "bold"),
    plot.title = element_blank()   # remove title
  )

```

<br>


<br><br>

## Beta Diversity

```{r beta_diversity_taxonomy, results="hide"}

set.seed(12345)

pst_Ab <- subset_samples(noneg_tax, Diet == "Ab")
pst_No <- subset_samples(noneg_tax, Diet == "No")

#  Transform sample counts
pst_Ab <- transform_sample_counts(pst_Ab, function(x) x / sum(x))
pst_No <- transform_sample_counts(pst_No, function(x) x / sum(x))

# CLR transform phyloseq objects at Phylum level
beta2_Ab <- pst_Ab %>%
  tax_transform(trans = "clr", rank = "Phylum")

beta2_No <- pst_No %>%
  tax_transform(trans = "clr", rank = "Phylum")

# generate distance matrix
psdist_Ab <- phyloseq::distance(beta2_Ab, method = "euclidean")
psdist_No <- phyloseq::distance(beta2_No, method = "euclidean")

```


**Beta-diversity of the Ab group**

Taxonomy

```{r beta-div_tax_renaming}

# renaming the tax table to remove the p__
library(microbiomeutilities)
otu_tab <- microbiome::abundances(noneg_tax)
tax_tab <- phyloseq::tax_table(noneg_tax)
tax_tab <- phyloseq::tax_table(noneg_tax)
tax_table(noneg_tax)[, colnames(tax_table(noneg_tax))] <- gsub(tax_table(noneg_tax)[, colnames(tax_table(noneg_tax))],     pattern = "[a-z]__", replacement = "")


# ADONIS test for Age_wk within the 'Ab' group
adonis2(psdist_Ab ~ sample_data(beta2_Ab)$Age_wk, permutations = 10000) 
#         Df SumOfSqs      R2     F Pr(>F)
# Model     5   199.84 0.15803 1.1731 0.3271
# Residual 25  1064.72 0.84197             
# Total    29  1264.55 1.00000


# plotting on the next chunk


```

<br>

**Beta-diversity of the No group**

Taxonomy

```{r beta-div_tax}

# renaming the tax table to remove the p__
library(microbiomeutilities)
otu_tab <- microbiome::abundances(noneg_tax)
tax_tab <- phyloseq::tax_table(noneg_tax)
tax_tab <- phyloseq::tax_table(noneg_tax)
tax_table(noneg_tax)[, colnames(tax_table(noneg_tax))] <- gsub(tax_table(noneg_tax)[, colnames(tax_table(noneg_tax))],     pattern = "[a-z]__", replacement = "")

# ADONIS test for Age_wk within the 'No' group
adonis2(psdist_No ~ sample_data(beta2_No)$Age_wk, permutations = 10000) 
#          Df   SumOfSqs   R2      F    Pr(>F)
# Model     4     146.24 0.12909 0.7411 0.5978
# Residual 20    986.58 0.87091              
# Total    24    1132.82 1.00000 

# no pairwise comparison since ns


library(dplyr)
library(viridis)
library(ggnewscale)


#install.packages("scico")   # run once
library(scico)

age_palette <- setNames(
  scico::scico(8, palette = "vikO"),  
  c("1","5","7","9","11")
)

# Transform sample counts
pst <- transform_sample_counts(noneg_tax, function(x) x / sum(x))

# CLR transform phyloseq objects at Broadclass level
beta <- pst %>%
  tax_transform(trans = "clr", rank = "Phylum")

# Factor setup
sample_data(beta)$Diet <- recode(sample_data(beta)$Diet,
                                 "Ab" = "MR+A",
                                 "No" = "MR")
                                 
sample_data(beta)$Diet <- factor(sample_data(beta)$Diet, levels = c("MR+A", "MR"))

sample_data(beta)$Age_wk <- factor(sample_data(beta)$Age_wk,
                                   levels = c("1","5","7","9","11"))

# Palettes
treatment_colors <- c("MR+A" = "#B22222", "MR" = "#377EB8")

shape_vals <- c("MR+A" = 16, "MR" = 17)

# Ordination
ord_all <- beta %>% ord_calc(method = "PCA")

# Plot
p <- ord_all %>%
  ord_plot(aes_shape = FALSE, points_show = FALSE) +
  geom_point(aes(color = Age_wk, shape = Diet),
             size = 3.2, alpha = 2, stroke = 0) +
  scale_shape_manual(values = shape_vals, name = "Diet") +
  # cividis gradient for Age (purple → yellow)
  scale_color_manual(
    values = age_palette,
    name = "Age (weeks)",
    guide = guide_legend(override.aes = list(size = 4))
  ) +
  ggnewscale::new_scale("color") +
  stat_ellipse(aes(color = Diet, group = Diet),
               level = 0.95, linewidth = 0.9, alpha = 0.9) +
  scale_color_manual(values = treatment_colors, name = "Diet") +
  labs(x = "PC1", y = "PC2") +
  theme_bw(base_size = 13) +
  theme(
    legend.position = "right",
    legend.box = "vertical",
    legend.title = element_text(size = 10, face = "bold"),
    legend.text = element_text(size = 9),
    legend.key.size = unit(0.45, "cm"),
    axis.text = element_text(size = 11, face = "bold"),
    axis.title = element_text(size = 12, face = "bold"),
    panel.grid = element_blank(),
    plot.margin = margin(5, 10, 5, 5),
    plot.title = element_blank()
  )

p

```


```{r composition-genus}

# renaming the tax table to remove the g__
otu_tab <- microbiome::abundances(noneg_tax)
tax_tab <- phyloseq::tax_table(noneg_tax)
tax_tab <- phyloseq::tax_table(noneg_tax)
tax_table(noneg_tax)[, colnames(tax_table(noneg_tax))] <- gsub(tax_table(noneg_tax)[, colnames(tax_table(noneg_tax))],     pattern = "[a-z]__", replacement = "")

library(scales)

# Factor setup 
sample_data(noneg_tax)$Age_wk <- factor(sample_data(noneg_tax)$Age_wk,
                                        levels = c("1","5","7","9","11"))
sample_data(noneg_tax)$Diet <- recode_factor(sample_data(noneg_tax)$Diet,
                                             "Ab" = "MR+A", "No" = "MR")


plot_genus <- function(ps_obj, diet_label, palette) {
  # --- subset manually ---
  meta <- sample_data(ps_obj)
  keep <- rownames(meta)[meta$Diet == diet_label]
  ps_sub <- prune_samples(keep, ps_obj)
  ps_sub <- prune_taxa(taxa_sums(ps_sub) > 0, ps_sub)
  stopifnot(nsamples(ps_sub) > 0)

  # merge samples by Age_wk
  merged <- phyloseq::merge_samples(ps_sub, "Age_wk")
  merged <- transform_sample_counts(merged, function(x) x / sum(x))

  # rebuild metadata
  sdf <- data.frame(
    Age_wk = sample_names(merged),
    Diet   = diet_label,
    row.names = sample_names(merged),
    stringsAsFactors = FALSE
  )
  sample_data(merged) <- phyloseq::sample_data(sdf)
  sample_data(merged)$Age_wk <- factor(sample_data(merged)$Age_wk,
                                       levels = c("1","5","7","9","11"))

  # plot
  p <- comp_barplot(
      merged,
      tax_level = "Genus",
      n_taxa = 14,
      palette = palette,
      transform = "compositional",
      bar_width = 0.975
    ) +
    theme_pubclean(base_size = 13) +
    theme(
      strip.background = element_rect(color = "black", fill = "white"),
      strip.text = element_text(size = 12, face = "bold"),
      axis.text.x = element_text(size = 12, face = "bold"),
      axis.text.y = element_text(size = 12, face = "bold"),
      axis.title.y = element_text(size = 10, face = "bold"),
      axis.title.x = element_text(size = 10, face = "bold"),
      legend.position = "bottom",
      legend.box = "horizontal",
      legend.text = element_text(size = 10),
      legend.title = element_text(size = 9, face = "bold"),
      plot.margin = margin(3, 3, 3, 3)
    ) +
    guides(fill = guide_legend(ncol = 5, keywidth = 0.6, keyheight = 0.5)) +
    scale_x_discrete(limits = c("1","5","7","9","11"), drop = FALSE) +
    facet_wrap(~Diet, scales = "free_x") +
    labs(y = "Relative abundance", x = "Age (weeks)")

  return(p)
}

# ---- run ----
custom_palette <- c("#450C54", "#6D597A", "#24868E", "#43AA8B", "#3E9A6F", "#90BE6D",
"#F9C74F", "#F3722C", "#D95F02", "#F9844A", "#F94144", "#B56576",
"#577590", "#264653", "#999949"
)

p_MRA <- plot_genus(noneg_tax, "MR+A", custom_palette)  # or "MR+A" if recoded
p_MR  <- plot_genus(noneg_tax, "MR", custom_palette)  # or "MR" if recoded

final_plot <- ggarrange(p_MRA, p_MR, nrow = 1, common.legend = TRUE, legend = "bottom")
final_plot

```

``` {r taxa_abund-prev}

library(dplyr)
library(microbiome)

# Transform to relative abundance
rel_abund <- microbiome::transform(noneg_tax, "compositional")

# Extract abundance and taxonomy
otu_tab <- as.data.frame(otu_table(rel_abund))
tax_tab <- as.data.frame(tax_table(rel_abund))

# Add Genus to abundance matrix
otu_tab$Genus <- tax_tab$Genus

# Reshape into long format for accuracy
library(tidyr)
otu_long <- otu_tab %>%
  tibble::rownames_to_column("OTU") %>%
  pivot_longer(-c(OTU, Genus), names_to = "Sample", values_to = "RelAbundance")

# Compute mean relative abundance per Genus
mean_abund_genus <- otu_long %>%
  group_by(Genus) %>%
  summarise(Mean_RelAbundance = mean(RelAbundance, na.rm = TRUE) * 100) %>%
  arrange(desc(Mean_RelAbundance))

# Get top 20
top20_genus <- head(mean_abund_genus, 20)

top20_genus

```

``` {r }
# ===========================================================
# SAVE REPRODUCIBILITY TABLES FOR GitHub
# ===========================================================

# Create folder if not exists
dir.create("taxonomic", showWarnings = FALSE)

# -------------------------------
# 1. Save alpha diversity tables
# -------------------------------

# Raw alpha
write.csv(alpha, "taxonomic/alpha_raw.csv", row.names = FALSE)

# Clean alpha (outliers removed)
write.csv(alpha_clean, "taxonomic/alpha_clean.csv", row.names = FALSE)

# Summary stats
write.csv(sum_Shannon, "taxonomic/sum_Shannon.csv", row.names = FALSE)
write.csv(sum_Observed, "taxonomic/sum_Observed.csv", row.names = FALSE)

lmm1_anova <- anova(lmm1)
shannon_lmm_df <- as.data.frame(lmm1_anova)
shannon_lmm_df$Term <- rownames(shannon_lmm_df)
write.csv(shannon_lmm_df, "lmm_Shannon.csv", row.names = FALSE)

lmm_obs_aov <- anova(lmm_obs)
observed_lmm_df <- as.data.frame(lmm_obs_aov)
observed_lmm_df$Term <- rownames(observed_lmm_df)
write.csv(observed_lmm_df, "lmm_Observed.csv", row.names = FALSE)


# Outlier IDs
write.table(outlier_ids, "taxonomic/outlier_ids.txt",
            row.names = FALSE, col.names = FALSE, quote = FALSE)


# -------------------------------
# 2. Save beta-diversity inputs
# -------------------------------

# Distance matrices
saveRDS(psdist_Ab, "taxonomic/psdist_Ab.rds")
saveRDS(psdist_No, "taxonomic/psdist_No.rds")

# ADONIS results
adonis_Ab <- adonis2(psdist_Ab ~ sample_data(beta2_Ab)$Age_wk, permutations = 10000)
adonis_No <- adonis2(psdist_No ~ sample_data(beta2_No)$Age_wk, permutations = 10000)

write.csv(as.data.frame(adonis_Ab), "taxonomic/adonis_Ab.csv")
write.csv(as.data.frame(adonis_No), "taxonomic/adonis_No.csv")


# -------------------------------
# 3. Save decontam results
# -------------------------------

write.csv(contamdf.prev, "taxonomic/decontam_results.csv")
write.table(contams_tax, "taxonomic/list_of_contaminants.txt",
            quote = FALSE, row.names = FALSE, col.names = FALSE)


# -------------------------------
# 4. Save taxonomic abundance tables
# -------------------------------

# Relative abundance table (compositional)
write.csv(otu_tab, "taxonomic/otu_table_rel_abund.csv")

# Long format relative abundance
write.csv(otu_long, "taxonomic/otu_table_rel_abund_long.csv", row.names = FALSE)

# Tax table used in that section
write.csv(tax_tab, "taxonomic/tax_table_cleaned.csv")


# -------------------------------
# 5. Save top taxa tables
# -------------------------------

write.csv(mean_abund_genus, "taxonomic/mean_abundance_genus.csv", row.names = FALSE)
write.csv(top20_genus, "taxonomic/top20_genus.csv", row.names = FALSE)


# -------------------------------
# 6. Save CLR-transformed tables for ordination
# -------------------------------

# Abundant CLR (Already in beta)
clr_Ab <- abundances(beta2_Ab)
clr_No <- abundances(beta2_No)

write.csv(clr_Ab, "taxonomic/clr_abundance_Ab.csv")
write.csv(clr_No, "taxonomic/clr_abundance_No.csv")


# -------------------------------
# 7. Save final phyloseq object (clean)
# -------------------------------

saveRDS(noneg_tax, "taxonomic/noneg_tax_clean.rds")

# ===========================================================
# DONE — all reproducibility tables saved to /taxonomic/
# ===========================================================
```






