---
title: "AMC-AMR_Stats"
output:
  html_document: default
  pdf_document: default
date: "2025-10-24"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  warning = FALSE,
  message = FALSE,
  fig.align = "center",
  fig.width = 6,
  fig.height = 4,
  out.width = "80%"
)



```

<br>

# AMR++

Antimicrobial Resistance Genes Analysis

```{r amrplusplus_analysis, results="hide"}
library(BiocManager)
library(tidyverse)
library(phyloseq)
library(microViz)
library(tidylog)
library(ggpubr)
library(vegan)
library(readxl)
library(dunn.test)
library(decontam)
library(tibble)
library(rstatix)
library(magrittr)
library(pairwiseAdonis)
library(paletteer)
library(speedyseq)

# Load phyloseq object
rawps_amr <- readRDS("data/rawps_amr.rds")

# Filter genes that require SNP Confirmation
psfilt <- rawps_amr %>% tax_select(tax_list = "SNP", strict_matches = FALSE, deselect = TRUE)

```

```{r decontam}

library(decontam)

# Decontamination using decontam package
ps2 <- psfilt %>%
  ps_mutate(
    SampleBinary = if_else(str_detect(Trt, "Negative"), true = "Control", false = "Sample") # ignore positive 
  )

# Identify contaminants using prevalence method
sample_data(ps2)$is.neg <- sample_data(ps2)$SampleBinary == "Control"
contamdf.prev <- isContaminant(ps2, method="prevalence", neg="is.neg", threshold = 0.5)

# Create phyloseq object of only contaminants in the controls
con <- rownames(contamdf.prev[contamdf.prev$contaminant == "TRUE", ])
conps <- prune_taxa(rownames(ps2@tax_table) %in% con, ps2)
negps <- subset_samples(conps, is.neg == "TRUE")

# Extract list of contaminants
contams <- rownames(contamdf.prev[contamdf.prev$contaminant == "TRUE", ])

# Display list of contaminants
cat("List of contaminants:\n")
print(contams) # 55

# Remove contaminant sequences
noncontam <- prune_taxa(!rownames(ps2@tax_table) %in% contams, ps2)

# Filter controls
controls <- ps2 %>%
  ps_filter(SampleBinary == "Control", .keep_all_taxa = TRUE)

controls <- sample_data(controls)$Project_ID 

# Remove all controls
noneg <- ps_filter(noncontam, !Project_ID %in% controls)
factor(sample_data(noneg)$Trt)

# Remove positive control
noneg <- ps_filter(noneg, Trt != "PositiveControl")
factor(sample_data(noneg)$Trt)

# Remove weeks 3, 13 and 15
noneg <- ps_filter(noneg, !Age_wk %in% c("3", "13", "15"))

# Prevalence-based filtering for negonly decontam
sort(phyloseq::sample_sums(noneg))
(noneg <- phyloseq::subset_samples(noneg, phyloseq::sample_sums(noneg) > 1))
(noneg <- phyloseq::prune_taxa(phyloseq::taxa_sums(noneg) > 0, noneg))
noneg <- noneg %>% tax_filter(
  min_prevalence = 0.01,
  min_total_abundance = 0.0001)

sample_data(noneg)

```

## Alpha Diversity

```{r alpha_diversity-summary"}

# Get alpha diversity metrics
sampdf <- sample_data(noneg) %>%
  data.frame() %>%
  rownames_to_column(var = "ID")

# Calculate alpha diversity
alpha <- estimate_richness(noneg, measures = c("Shannon", "Observed")) %>%
  rownames_to_column(var = "ID") %>%
  merge(sampdf, by = "ID")

# Get summary statistics for Shannon diversity
sum_Shannon <- alpha %>%
  group_by(Trt, Age_wk) %>%
  get_summary_stats(Shannon, type = "mean_sd")

# Print summary statistics
print(sum_Shannon)

# Get summary statistics for Observed diversity
sum_Observed <- alpha %>%
  group_by(Trt, Age_wk) %>%
  get_summary_stats(Observed, type = "mean_sd")

# Print summary statistics
print(sum_Observed)

```

```{r alpha_diversity-outliers, results="hide"}

## Outliers 

# Checking Outliers for Shannon Diversity
st <- alpha %>% get_summary_stats(Shannon, type = "mean_sd") %>%
  mutate(sdx3 = sd * 3,
         outhi = mean + sdx3,
         outlo = mean - sdx3)

shannon_lower_outliers <- alpha$ID[alpha$Shannon < st$outlo]
shannon_upper_outliers <- alpha$ID[alpha$Shannon > st$outhi]

print(shannon_lower_outliers) # 0
print(shannon_upper_outliers) # 0
print(st$sd) # 0.801

# Checking Outliers for Observed Diversity
sto <- alpha %>% get_summary_stats(Observed, type = "mean_sd") %>%
  mutate(sdx3 = sd * 3,
         outhi = mean + sdx3,
         outlo = mean - sdx3)

observed_lower_outliers <- alpha$ID[alpha$Observed < sto$outlo]
observed_upper_outliers <- alpha$ID[alpha$Observed > sto$outhi]

print(observed_lower_outliers) # 0
print(observed_upper_outliers) # 0
print(sto$sd) # 206.364

```

<br>

Shapiro-Wilk Test for Normality

```{r normality_tests}

# Statistical Tests

# Shapiro-Wilk Test for Normality
shapiro_shannon <- shapiro.test(alpha$Shannon)
shapiro_observed <- shapiro.test(alpha$Observed)

shapiro_shannon # W = 0.75961, p-value = 4.095e-08
shapiro_observed # W = 0.85721, p-value = 1.114e-05

```

<br>

[Wilcox Test for Diet Effect]{.underline}

Mann-Whitney and BH adjustment

**Shannon Index**

AMR Genes

```{r Shannon}

## Doing first within Ab and then No - Shannon 
alpha_Ab <- subset(alpha, Diet == "Ab")
alpha_No <- subset(alpha, Diet == "No")

alpha_Ab$Age_wk <- as.numeric(alpha_Ab$Age_wk)
alpha_No$Age_wk <- as.numeric(alpha_No$Age_wk)

kruskal.test(Shannon ~ Age_wk, data = alpha_Ab) # Kruskal-Wallis chi-squared = 1.8624, df = 4, p-value = 0.7611
kruskal.test(Shannon ~ Age_wk, data = alpha_No) # Kruskal-Wallis chi-squared = 1.3218, df = 4, p-value = 0.8577
# not necessary to run Dunn bc ns

results_Shannon <- list()

# Loop through each week and perform the Mann-Whitney U test
for (week in unique(alpha$Age_wk)) {
  data_week <- subset(alpha, Age_wk == week)

  # Mann-Whitney for Shannon
  test_result_Shannon <- wilcox.test(Shannon ~ Trt, data = data_week)
  results_Shannon[[paste0("Week_", week)]] <- test_result_Shannon$p.value
}

# Convert results to data frames 
results_df_Shannon <- data.frame(Week = names(results_Shannon), P_Value = unlist(results_Shannon))

# Adjust p-values to BH method
results_df_Shannon$Adjusted_P_Value_BH <- p.adjust(results_df_Shannon$P_Value, method = "BH")

# Add a column for the groups compared
results_df_Shannon$Groups_Compared <- "Ab vs No"

# plotting merged alphas on next chunk 

```

<br>


**Observed Index**

AMR Genes

```{r Observed, fig.height=6, fig.width=8}

## Doing first within Ab and then No - Shannon 
alpha_Ab <- subset(alpha, Diet == "Ab")
alpha_No <- subset(alpha, Diet == "No")

alpha_Ab$Age_wk <- as.numeric(alpha_Ab$Age_wk)
alpha_No$Age_wk <- as.numeric(alpha_No$Age_wk)

kruskal.test(Observed ~ Age_wk, data = alpha_Ab) # Kruskal-Wallis chi-squared = 1.5372, df = 4, p-value = 0.82
kruskal.test(Observed ~ Age_wk, data = alpha_No) # Kruskal-Wallis chi-squared = 0.82064, df = 4, p-value = 0.9357


## Comparing Ab and No regarding same week

results_Observed <- list()

# Loop through each week and perform the Mann-Whitney U test
for (week in unique(alpha$Age_wk)) {
  data_week <- subset(alpha, Age_wk == week)

  # Mann-Whitney for Observed
  test_result_Observed <- wilcox.test(Observed ~ Trt, data = data_week)
  results_Observed[[paste0("Week_", week)]] <- test_result_Observed$p.value
}
  
# Convert results to data frames 
results_df_Observed <- data.frame(Week = names(results_Observed), P_Value = unlist(results_Observed))

# Adjust p-values to BH method
results_df_Observed$Adjusted_P_Value_BH<- p.adjust(results_df_Observed$P_Value, method = "BH")

# Add a column for the groups compared
results_df_Observed$Groups_Compared <- "Ab vs No"

# Print the results
print(results_df_Observed)

# plotting
alpha$Age_wk <- factor(alpha$Age_wk, levels = c(1, 5, 7, 9, 11))

alpha$Diet <- recode_factor(alpha$Diet,
                            "Ab" = "MR+A",
                            "No" = "MR")

treatment_colors <- c("MR+A" = "#450c54", "MR" = "#24868E")

ggline(alpha, x = "Age_wk", y = "Shannon",
       add = "mean_sd", error.plot = "errorbar",
       color = "Diet", group = "Diet",
       palette = treatment_colors,
       size = 1.0, point.size = 1.5, alpha = 0.8) +
  scale_x_discrete(labels = c("1","5","7","9","11")) +
  labs(x = "Age (weeks)", y = "Shannon Diversity") +
  theme_classic(base_size = 12) +
  theme(
    legend.position = "right",
    axis.title = element_text(size = 12, face = "bold"),
    plot.title = element_blank()   # remove title
  )

ggline(alpha, x = "Age_wk", y = "Observed",
       add = "mean_sd", error.plot = "errorbar",
       color = "Diet", group = "Diet",
       palette = treatment_colors,
       size = 1.0, point.size = 1.5, alpha = 0.8) +
  scale_x_discrete(labels = c("1","5","7","9","11")) +
  labs(x = "Age (weeks)", y = "Observed Diversity") +
  theme_classic(base_size = 12) +
  theme(
    legend.position = "right",
    axis.title = element_text(size = 12, face = "bold"),
    plot.title = element_blank()   # remove title
  )

```

<br>

[Wilcoxon Test for Diet Effect]{.underline}

**Shannon and Observed Indexes**

AMR Genes

```{r}

#Test treatment 
wilcox.test(Shannon ~ Diet, data = alpha) # W = 399, p-value = 0.6935
wilcox.test(Observed ~ Diet, data = alpha) # W = 388.5, p-value = 0.8261

```

<br><br>

## Beta Diversity

```{r beta_diversity, results="hide"}

set.seed(12345)

pst_Ab <- subset_samples(noneg, Diet == "Ab")
pst_No <- subset_samples(noneg, Diet == "No")

#  Transform sample counts
pst_Ab <- transform_sample_counts(pst_Ab, function(x) x / sum(x))
pst_No <- transform_sample_counts(pst_No, function(x) x / sum(x))

# CLR transform phyloseq objects at Broadclass level
beta1_Ab <- pst_Ab %>%
  tax_transform(trans = "clr", rank = "Broadclass")

beta1_No <- pst_No %>%
  tax_transform(trans = "clr", rank = "Broadclass")

# generate distance matrix
psdist_Ab <- phyloseq::distance(beta1_Ab, method = "euclidean")
psdist_No <- phyloseq::distance(beta1_No, method = "euclidean")

```


**Beta-diversity of the Ab group**

AMR Genes

```{r beta_div-Ab}

# ADONIS test for Age_wk within the 'Ab' group
adonis2(psdist_Ab ~ sample_data(beta1_Ab)$Age_wk, permutations = 10000) 

#          Df SumOfSqs      R2      F Pr(>F)
# Model     4   0.6555 0.0881 0.6038 0.8025
# Residual 25  6.7848 0.9119              
# Total    29  7.4402 1.00000  

# plotting merged betas on the next chunk

```

<br>

**Beta-diversity of the No group**

AMR Genes

```{r beta_div-No}

# ADONIS test for Age_wk within the 'No' group
adonis2(psdist_No ~ sample_data(beta1_No)$Age_wk, permutations = 10000)

#          Df SumOfSqs      R2      F Pr(>F)  
# Model     4   0.39113 0.1285 0.7372 0.6561 
# Residual 20   2.65278 0.8715                
# Total    24   3.04392 1.00000   

# no pairwise comparison since ns

## for plotting 

#  Transform sample counts
pst <- transform_sample_counts(noneg, function(x) x / sum(x))

# CLR transform phyloseq objects at Broadclass level
beta <- pst %>%
  tax_transform(trans = "clr", rank = "Broadclass")


sample_data(beta)$Diet <- recode(sample_data(beta)$Diet,
                                     "Ab" = "MR+A",
                                     "No" = "MR")

sample_data(beta)$Diet <- factor(sample_data(beta)$Diet, levels = c("MR+A", "MR"))
sample_data(beta)$Age_wk <- factor(sample_data(beta)$Age_wk,
                                       levels = c("1","5","7","9","11"))

# 2️⃣ Palettes
treatment_colors <- c("MR+A" = "#450c54", "MR" = "#24868E")
age_levels  <- levels(sample_data(beta)$Age_wk)
age_palette <- setNames(viridis::viridis(length(age_levels), option = "G"), age_levels)
shape_vals  <- c("MR+A" = 16, "MR" = 17)

# 3️⃣ Ordination
ord_all <- beta %>% ord_calc(method = "PCA")

# 4️⃣ Plot
p <- ord_all %>%
  ord_plot(aes_shape = FALSE, points_show = FALSE) +
  geom_point(aes(color = Age_wk, shape = Diet),
             size = 3.2, alpha = 2, stroke = 0) +
  scale_shape_manual(values = c("MR+A" = 16, "MR" = 17), name = "Diet") +
  scale_color_manual(values = viridis::viridis(8, option = "G"),
                     name = "Age (weeks)",
                     guide = guide_legend(override.aes = list(size = 4))) +
  ggnewscale::new_scale("color") +
  stat_ellipse(aes(color = Diet, group = Diet),
               level = 0.95, linewidth = 0.9, alpha = 0.9) +
  scale_color_manual(values = c("MR+A" = "#450c54", "MR" = "#24868E"),
                     name = "Diet") +
  labs(x = "PC1", y = "PC2", caption = NULL) +
  theme_bw(base_size = 13) +
  theme(
    legend.position = "right",
    legend.box = "vertical",
    legend.title = element_text(size = 11, face = "bold"),
    legend.text = element_text(size = 10),
    axis.text = element_text(size = 11, face = "bold"),
    axis.title = element_text(size = 12, face = "bold"),
    text = element_text(size = 13),
    panel.grid = element_blank(),
    plot.margin = margin(5, 10, 5, 5),
    plot.title = element_blank()
  )


p

```

<br><br>

## Compositional plots

Broad class and class

```{r comp_amr}
###############################################################################
# LOAD LIBRARIES
###############################################################################
library(phyloseq)
library(microbiome)
library(tidyverse)
library(readr)
library(viridisLite)
library(ComplexHeatmap)
library(circlize)
library(grid)


###############################################################################
# 0) REMOVE CONTROLS FIRST
###############################################################################
noneg <- prune_samples(
  !sample_names(noneg) %in% c("AMC550", "AMC551", "AMC552"),
  noneg
)


###############################################################################
# 1) AGGREGATE TO AMR CLASS
###############################################################################
psrel   <- microbiome::transform(noneg, "compositional")
psclass <- tax_glom(psrel, taxrank = "Class")

mat <- as.data.frame(otu_table(psclass))
class_names <- as.character(tax_table(psclass)[, "Class"])
rownames(mat) <- class_names

mat <- mat %>%
  rownames_to_column("Class") %>%
  group_by(Class) %>%
  summarise(across(everything(), mean)) %>%
  column_to_rownames("Class")

z3 <- scale(as.matrix(mat))
z3 <- as.matrix(z3)


###############################################################################
# 2) METADATA
###############################################################################
AMC_metadata <- read_csv(
  "~/OneDrive - The Pennsylvania State University/AMC_samples/AMC-metadata.csv",
  col_types = cols(Project_ID = col_character())
)

AMC_metadata <- AMC_metadata %>%
  mutate(across(where(is.character), ~na_if(trimws(.), "")))

keep <- colnames(z3)

met <- AMC_metadata %>%
  filter(Project_ID %in% keep) %>%
  filter(!Age_wk %in% c(3, 13, 15))

met$Treatment <- case_when(
  str_detect(met$Diet, "Ab") ~ "MR+A",
  str_detect(met$Diet, "No") ~ "MR",
  TRUE ~ met$Diet
)

met <- met[match(keep, met$Project_ID), ]
stopifnot(identical(met$Project_ID, keep))

met2 <- met %>% select(Treatment, Age_wk)
rownames(met2) <- met$Project_ID
met2$Treatment <- factor(met2$Treatment, levels = c("MR+A", "MR"))
met2$Age       <- factor(as.numeric(met2$Age_wk))


###############################################################################
# 3) RESISTANCE CLASS (Broadclass)
###############################################################################
merge3 <- as.data.frame(tax_table(psclass)) %>%
  select(Class, Broadclass)

rownames(merge3) <- merge3$Class
merge3 <- merge3[rownames(z3), , drop = FALSE]
stopifnot(identical(rownames(merge3), rownames(z3)))

merge3$Mean <- rowMeans(mat)[merge3$Class]

top5_df <- merge3 %>%
  group_by(Broadclass) %>%
  arrange(desc(Mean)) %>%
  slice_head(n = 5)

top5 <- top5_df$Class

z3_sub <- z3[top5, , drop = FALSE]
merge3_sub <- merge3[top5, , drop = FALSE]
stopifnot(identical(rownames(z3_sub), rownames(merge3_sub)))

row_split <- merge3_sub$Broadclass


###############################################################################
# 4) ANNOTATIONS
###############################################################################
treatment_colors <- c("MR+A" = "#450c54", "MR" = "#24868E")

age_palette <- setNames(
  viridis(length(levels(met2$Age)), option = "G"),
  levels(met2$Age)
)

ha_col <- HeatmapAnnotation(
  df = met2[, c("Treatment", "Age")],
  col = list(
    Treatment = treatment_colors,
    Age       = age_palette
  )
)

resistance_colors <- c(
  "Biocides"       = "#552F7A",
  "Drugs"          = "#7C5F98",
  "Metals"         = "#B09FC1",
  "Multi-compound" = "#CABED6"
)

ha_row <- rowAnnotation(
  Resistance = merge3_sub$Broadclass,
  col = list(Resistance = resistance_colors),
  show_annotation_name = FALSE,
  annotation_name_gp = gpar(fontsize = 0)
)


###############################################################################
# 5) ORDER COLUMNS WITHOUT LABELS
###############################################################################
ord <- order(met2$Treatment, met2$Age)
z3_sub <- z3_sub[, ord, drop = FALSE]
met2   <- met2[ord, , drop = FALSE]


###############################################################################
# 6) FINAL HEATMAP — NO LEFT TEXT, NO LABELS, SMALL GROUP GAPS
###############################################################################
ht <- Heatmap(
  z3_sub,
  name            = "Z-score",
  top_annotation  = ha_col,
  left_annotation = ha_row,

  row_split       = row_split,
  row_title       = NULL,                 # hide split labels
  row_title_gp    = gpar(fontsize = 0),   # fully invisible
  row_gap         = unit(1.2, "mm"),      # slightly smaller spacing

  cluster_rows    = TRUE,
  cluster_columns = TRUE,
  show_column_names = FALSE,

  col = colorRamp2(c(-2, 0, 2), viridis(3, option = "G")),
  row_names_gp = gpar(fontsize = 10),
  border = TRUE,
  heatmap_legend_param = list(title = "Z-score")
)

draw(ht)


```
``` {r }
# ========================================================================

# SAVE ALL PROCESSED AMR OUTPUTS FOR REPRODUCIBILITY

# ========================================================================


## 1) Decontam outputs

write.csv(contamdf.prev,           "amr/decontam_results.csv", row.names = TRUE)
writeLines(contams,                "amr/list_of_contaminants.txt")

## 2) Cleaned phyloseq objects

saveRDS(noneg,                     "amr/noneg_amr_clean.rds")
saveRDS(psdist_Ab,                 "amr/psdist_Ab.rds")
saveRDS(psdist_No,                 "amr/psdist_No.rds")

## 3) Alpha diversity tables

write.csv(alpha,                   "amr/alpha_raw.csv", row.names = FALSE)
write.csv(sum_Shannon,             "amr/sum_Shannon.csv", row.names = FALSE)
write.csv(sum_Observed,            "amr/sum_Observed.csv", row.names = FALSE)
write.csv(results_df_Shannon,      "amr/wilcox_Shannon_byweek.csv", row.names = FALSE)
write.csv(results_df_Observed,     "amr/wilcox_Observed_byweek.csv", row.names = FALSE)

## 4) Outliers

writeLines(shannon_upper_outliers, "amr/outlier_Shannon_upper.txt")
writeLines(observed_upper_outliers,"amr/outlier_Observed_upper.txt")

## 5) Beta diversity / PERMANOVA

write.csv(adonis2(psdist_Ab ~ sample_data(beta1_Ab)$Age_wk),
"amr/beta_PERMANOVA_Ab.csv")
write.csv(adonis2(psdist_No ~ sample_data(beta1_No)$Age_wk),
"amr/beta_PERMANOVA_No.csv")

## 6) CLR abundance matrices for both groups

clr_abundance_Ab <- as.matrix(otu_table(beta1_Ab))
clr_abundance_No <- as.matrix(otu_table(beta1_No))

write.csv(clr_abundance_Ab,        "amr/clr_abundance_Ab.csv")
write.csv(clr_abundance_No,        "amr/clr_abundance_No.csv")

## 7) Resistance class heatmap matrices

write.csv(z3,                      "amr/zscore_matrix_class.csv")
write.csv(z3_sub,                  "amr/zscore_matrix_class_top5.csv")

## 8) Tax tables

write.csv(as.data.frame(tax_table(noneg)),
"amr/tax_table_cleaned.csv")

## 9) Composition summaries

write.csv(mat,                     "amr/otu_table_rel_abund.csv")
write.csv(merge3,                  "amr/AMR_class_abundance_table.csv")
write.csv(top5_df,                 "amr/AMR_class_top5.csv")

cat("✓ All AMR output files saved in /amr folder\n")

```
